feat: Code Language Selector

mitsuhiko opened the pull request:
This adds improvements to code blocks and language selectors. Sibling code blocks automatically fold into a tab box unless separated by other content (or `<Break/>`):

* `{filename: name-of-file.ext}`: Adds a filename overlay to code block
* `{forceTabBar}`: Forces the tab bar even if there is only one language in the block
* `{title: Name}`: Overrides the default title (language name). Can be use to make two blocks of one language. Eg: `React` and `Vue`.

Code blocks automatically get sorted so the order in which they are appearing in the source file is irrelevant.

Example: https://github.com/getsentry/develop/blob/1a2911120e68f75f0bea5f0a04799deac1466695/src/docs/docs-components.mdx#L9-L31

Turns into: https://develop-bakzlepnj.sentry.dev/docs-components/

![image](https://user-images.githubusercontent.com/7396/81204145-1e6ed580-8fc9-11ea-8add-dcb57398b6a5.png)

mitsuhiko commented:
Last commit is just to demo it on vercel.

matejminar commented on src/components/codeTabs.js:
```diff
@@ -0,0 +1,85 @@
+        className={isSelected ? "active" : ""}
+        onClick={() => {
+          setSharedSelection(choice);
+          setLocalSelection(choice);
```
I believe you don't have to call `setLocalSelection `.

CodeTabs component will be rerendered because of change in context caused by `setSharedSelection ` and therefore localSelection will be updated too.

mitsuhiko replied:
But that will require another re-render, no?

matejminar replied:
o yeah, let's leave as is 👍

HazAT commented:
Can we maybe have an experimentation page with all components on it instead of having this there?

mitsuhiko commented:
@HazAT yeah, that's just a small temporary hack.

dcramer commented:
Some carry over from Slack...

I worry about this with e.g. 15 code samples, so if we're going to keep our onboarding with the mega-language selector, I'd say we avoid this pattern for that and come up with something that's easier to create consistency behind. That said, I'm not sure our goal is to keep that style of onboarding guide (vs just saying "go to this language docs"). This still is valuable for things like "show obj-c vs swift".

Few other things:

1) Could we put a basic test in to make sure we don't fuck this up in the future? https://www.gatsbyjs.org/docs/testing/
2) We could battle test this in the Dev Env guide (zsh vs bash examples)
3) This week I will try to get a third navigation tier into the sidebar (along with some other nav improvements), and maybe we can (under Documentation Guide) have an examples section, or specific callouts to how some of the tech works there.

mitsuhiko commented:
Added some very basic tests for the markdown part. Need to still figure out how to actually test the react components that it throws into the code.

I moved the components into `docs-components`. So far not linked from the sidebar.

HazAT commented:
@dcramer I was talking with Fiona and we basically want that the getting started page of the docs is equal to what we have as the onboarding wizard in Sentry right now.

No language selector on getting started, rather select exactly what platform/sdk you care about and we guide you to the page.
